markdown: Support inline images using the `![alt text](url)` syntax.

karlstolley opened the pull request:
This PR enables native Markdown support for inline images.

The following work remains here:

- [x] Address outstanding comments from the original PR, #34732
- [x] Move the commit for inserting the `![]()` syntax to the end of the commit list, so that the bulk of the logic here can be easily deployed to CZO for testing
- [x] Revise postprocess and message-list-hover logic to handle animated inline images
- [x] Add inline-images postprocess tests to cover animate-on-hover behavior
- [x] Double-check comments in `markdown/__init.py__` to make sure they don't reference functions on the frontend that have since been renamed
- [x] Modify Turndown to correctly handle quote + reply behavior on messages that include the inline-image syntax
- [x] Document the changes here, currently presented in a standalone PR: #36278
- [x] Add screenshots showing inline, some legacy images + replies, quote & replies
- [ ] Correct for non-image files (uploaded and remote) presented in the `![]()` syntax, and present them as links to the files in question: [CZO discussion](https://chat.zulip.org/#narrow/channel/3-backend/topic/inline-image.20corner.20cases/near/2278074)
- [ ] Possibly set the `has_image` flag on `self.zmd.zulip_message` in the `InlineImageProcessor` class, parallel to that action in `InlineInterestingLinkProcessor`: [CZO discussion](https://chat.zulip.org/#narrow/channel/3-backend/topic/inline-image.20corner.20cases/near/2278211)
- [x] Look at [deduplicating](https://github.com/zulip/zulip/pull/34732#discussion_r2395016995) the `zulip_specific_src_changes` function > unsure about this; might be a follow-up and something I open a discussion about in `#backend`
- [ ] The Changes section of the documentation commit must be updated with the correct Zulip version and feature levels when the changes from this PR are ready to ship.

<!-- Describe your pull request here.-->

Fixes: #28912

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

_Showing the original image presentation, the whole-message reply, and quote + reply:_

| Legacy images | Inline images | Tiny inline images | Me-message images | Inline gallery images|
| --- | --- | --- | --- |  --- |
| <img width="1400" height="2200" alt="legacy-image-replies" src="https://github.com/user-attachments/assets/e3ed06cf-27d1-4d1f-86e7-5de3813bae7d" /> | <img width="1400" height="2200" alt="inline-image-replies" src="https://github.com/user-attachments/assets/61182015-18a6-4b4a-b373-a2760bdf3d97" /> | <img width="1400" height="2200" alt="inline-tiny-image-replies" src="https://github.com/user-attachments/assets/be2a252b-7e2d-4115-b73b-41b1fc1e92f0" /> | <img width="1400" height="2200" alt="inline-me-message-image-replies" src="https://github.com/user-attachments/assets/410d978d-17e5-4c7f-9b1d-14b5656a7f2f" /> | <img width="1400" height="2200" alt="inline-gallery-image-replies" src="https://github.com/user-attachments/assets/df8995a9-41fe-4e22-b471-e5542544f9be" /> |

_Showing the expected behavior of loading spinners while inline images are thumbnailed:_

![inline-images-loading](https://github.com/user-attachments/assets/cb9f1f43-0d55-42b4-b047-b86640375d46)







<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

karlstolley committed:
api_docs: Introduce images in Markdown syntax.

Co-authored-by: Alex Vandiver <alexmv@zulip.com>

timabbott commented:
Added the one commit from #36278 here. It still needs **Changes** notes to be complete.

alya commented:
Hm, we may need to more aggressively put images in a gallery. Uploading 3 images in a row (without typing anything), I ended up with each image on a new line, and had to delete the line breaks between them to get a gallery. I think images on consecutive lines (no empty lines in between) should get displayed together.

Alternately, we could place consecutive uploads on the same line, but that's rather hard to read.

alya commented:
I'm seeing an outline on hover when an image is by itself, but not when it, e.g., has some text before it. Is that expected?

karlstolley commented:
In response to @alya's last round of feedback:

1. Providing a simple link for a non-image file sounds reasonable. There's a backend `url_to_a` function that would provide a link based on a URL, once the URL has been properly sanitized. I've added this as a TODO to the top of the PR.

2. I'm not able to reproduce a gallery problem when each file sits on its own line. Unless there's a blank line of space between them, those should all be treated as part of a single paragraph, and therefore a single gallery.

3. Yes, image previews don't show up in Drafts because of how we process images server-side. So while that isn't ideal behavior, it is at least currently expected.

4. In [#design > inline image presentation](https://chat.zulip.org/#narrow/channel/101-design/topic/inline.20image.20presentation/with/2258026), we decided to try omitting a border around images that sit in and amongst other text, and reserve the border only for gallery images. Probably that's something to evaluate further once this gets deployed.

timabbott commented on web/src/compose_paste.ts:
```diff
@@ -38,10 +38,25 @@ function image_to_zulip_markdown(
+        node.getAttribute("href") ??
+        "";
+    // We prioritize the alt attribute, which will be present on
+    // inline Markdown images
```
What's the motivation for preferring `title` vs `alt`? It sounds like the corner case you're thinking about is one where only `alt` is present.

timabbott replied:
``` diff
diff --git a/web/src/compose_paste.ts b/web/src/compose_paste.ts
index 880fef70cf..f8fb9b4201 100644
--- a/web/src/compose_paste.ts
+++ b/web/src/compose_paste.ts
@@ -48,8 +48,10 @@ function image_to_zulip_markdown(
         node.getAttribute("src") ??
         node.getAttribute("href") ??
         "";
-    // We prioritize the alt attribute, which will be present on
-    // inline Markdown images
+    // For the label, we prioritize the alt attribute, which will be
+    // present on both Zulip and third-party Markdown images, but use
+    // title as a fallback, since older-style Zulip image previews
+    // only set title.
     let alt = "";
     if (node.classList.contains("inline-image")) {
         alt = node.getAttribute("alt") ?? "";
```

timabbott commented on zerver/lib/markdown/__init__.py:
```diff
@@ -2123,6 +2123,80 @@ def handleMatch(  # type: ignore[override] # https://github.com/python/mypy/issu
+        self.zmd = zmd
+
+    def zulip_specific_src_changes(self, img: Element) -> None | Element:
+        # function partially copied from LinkInlineProcessor.zulip_specific_link_changes
```
I think it's probably not worth refactoring that at this time. It is pretty minor and the attributes are different.

timabbott commented on web/src/markdown.ts:
```diff
@@ -736,6 +736,10 @@ export function parse({
 
+    // Treat images as links, because only the backend markdown processor
+    // can handle them correctly.
+    renderer.image = renderer.link;
```
Dropped these two hunks from what I'm merging right now.

``` diff
diff --git a/web/src/markdown.ts b/web/src/markdown.ts
index 8d15207f93..2fa5fa2e35 100644
--- a/web/src/markdown.ts
+++ b/web/src/markdown.ts
@@ -736,10 +736,6 @@ export function parse({
         return "<br>\n";
     };
 
-    // Treat images as links, because only the backend markdown processor
-    // can handle them correctly.
-    renderer.image = renderer.link;
-
     function preprocess_code_blocks(src: string): string {
         return fenced_code.process_fenced_code(src);
     }
diff --git a/web/third/marked/lib/marked.d.cts b/web/third/marked/lib/marked.d.cts
index bd11c09f9b..e229afb892 100644
--- a/web/third/marked/lib/marked.d.cts
+++ b/web/third/marked/lib/marked.d.cts
@@ -10,7 +10,6 @@ declare class Renderer {
     code: (code: string) => string;
     link: (href: string, title: string, text: string) => string;
     br: () => string;
-    image: (href: string, title: string, text: string) => string;
 }
 
 declare namespace marked {
```

timabbott commented:
I opened #36822 to merge the several commits that don't involve actually generating the syntax. I can then review the backend a bit more tightly and then test-deploy it on chat.zulip.org.

karlstolley commented:
Noting here that we need to be sure to include the `.inline-image` class in functions like `message_has_image()` and test cases in `filter.test.cjs` adjacent the existing server-side `.message_inline_image` class, so as to avoid breaking the `has:image` filter.

See [#issues > has:image doesn't work. @ 💬](https://chat.zulip.org/#narrow/channel/9-issues/topic/has.3Aimage.20doesn't.20work.2E/near/2314411)

chrisbobbe reviewed (dismissed):
Thanks! Requesting small docs changes from discussion at [#api documentation > documenting inline images @ 💬](https://chat.zulip.org/#narrow/channel/412-api-documentation/topic/documenting.20inline.20images/near/2282569).

chrisbobbe commented on api_docs/message-formatting.md:
```diff
@@ -208,6 +227,18 @@ desired. For example:
+  data-original-content-type="image/png"
+  data-original-dimensions="1050x700"
+  data-original-src="/user_uploads/path/to/example.png"
+  src="/path/to/spinner.png">
```
Ah; I know it's a hardcoded spinner image URL, but perhaps should've guessed that it wasn't covered when you said (at [#api documentation > documenting inline images @ 💬](https://chat.zulip.org/#narrow/channel/412-api-documentation/topic/documenting.20inline.20images/near/2279225)):

> I think the current proposal that we not generate that output for anything but a thumbnailed image.

🙂 Still, if we're implicitly (or explicitly, as in my other review comment) telling clients they can expect a thumbnail URL in the normal, _non_-loading case, let's add a line to the doc here like:

> `src` is a hard-coded spinner image URL, and is not necessarily a thumbnail URL.

because it would be easy for clients to mistakenly expect a thumbnail URL here too.

alexmv commented on zerver/lib/thumbnail.py:
```diff
@@ -535,6 +537,111 @@ def get_transcoded_format(
 
+def process_inline_images_to_thumbnails(
+    image_tag: Tag | None,
+    image_src: str,
```
This doesn't look used?

alexmv commented on zerver/lib/thumbnail.py:
```diff
@@ -535,6 +537,111 @@ def get_transcoded_format(
+    image_link: Tag,
+) -> tuple[bool, str | None]:
+    changed = False
+    remaining_thumbnails_to_add = None
```
This variable is badly named -- it's only ever a single thumbnail, and we're only "adding" it in the sense that we happen to call `.add` with it on a set.

I'd call this `unthumbnailed_path_id` or `remaining_path_id`

alexmv committed:
thumbnail: Remove an unused parameter.

Introduced in e9f7e580723b.

alexmv committed:
thumbnail: Rename "inline_image" to be more specific.

alexmv committed:
thumbnail: Remove "remaining_thumbnails_to_add" variable.

This variable is confusingly named, since it only ever has one
thumbnail in it (as a path_id) and the thumbnail is not "added",
except to the set of un-thumbnailed paths.

Refactor the logic to simply return the tuples directly.

alexmv committed:
thumbnail: Remove an explicit cast().

alexmv committed:
thumbnail: Update comments and function names.

"Modern" and "traditional" have no meaning to people reading these
after they have been committed -- those concepts stem from the order in
which we happen to have written features, which is irrelevant to any
later readers.

Attempt to clarify link-based image thumbnails vs inline-image
thumbnails in comments; this is confused by the use of
`class="message_inline_image"` for the non-inline link-derived images.

alexmv committed:
thumbnail: Only support Markdown image syntax for uploaded images.

chrisbobbe commented on api_docs/message-formatting.md:
```diff
@@ -186,18 +186,53 @@ generate an image preview element with the following format:
+`/user_uploads/thumbnail/`).  If the image is *not* a thumbnail,
+clients should make a best-effort attempt to render the given `src`,
+in a manner which minimizes layout shift when the resource is loaded
+-- see the next section for a special case of this.
```
I'm not sure whether "guaranteed" in the first sentence should be read as covering historical messages (@gnprice, thoughts? 🙂). If so:

- It breaks for `data-original-content-type` before FL 336.
- It breaks for `data-original-dimensions` at FL 275.

I learned this from the helpful "Changes to image formatting" section. That section is linked just after this paragraph, so maybe no change is needed, and I'm just overinterpreting?

gnprice replied:
Just on the specific question: yeah, what I take as the purpose of this page is to document the content HTML that the server might send to clients, not only the content HTML that it might produce on new messages. So if something's said to be guaranteed, I'd expect that to apply to all existing messages on the server.

chrisbobbe reviewed (commented):
Thanks, this is very helpful! Small comment below.

chrisbobbe reviewed (commented):
Just one small comment, otherwise LGTM! 👍

timabbott commented:
I pushed back here with a couple changes:
- Addressed https://github.com/zulip/zulip/pull/36226#discussion_r2813778300 as described.
- Added an API feature level bump for this change being integrated, which includes the syntax being disabled for third-party images, and made sure the documentation clarified that explicitly.

timabbott commented:
With those tweaks, I'm merging this, aside from the one commit to make the web app start using this; I'll be putting that on a new PR that we test-deploy to chat.zulip.org shortly; I just want to protect `main` from this being heavily used by default until we've completed that end-to-end testing.

Opened #38005 for that new pull request.

timabbott commented:
CI is still failing, so I've moved a few early prep commits to #38007 just to shrink this, and will await clarification on what the story is with the tests and detail noted above.

alexmv commented on zerver/lib/thumbnail.py:
```diff
@@ -569,14 +566,13 @@ def process_inline_images_to_thumbnails(
-            # second image, the first image will have no placeholder.
-            pass
-        return changed, remaining_thumbnails_to_add
+            return True, path_id
```
A previous refactor commit in the stack had dropped a `changed = True` in this codepath -- so `return True, path_id` _was correct_.

I've fixed it back to `True`, and also removed the `test_inline_image` test which came in during 48fac910b5ddd3fbb0e644f22251ce18cdfbf0e7, which is no longer correct because `![]` now only works for uploaded images.

timabbott commented:
Fixed that last detail and merged, thanks for all the work on this @karlstolley, @alexmv, and @roanster007!

killermoehre commented:
@timabbott sorry for bringing this PR up again, but I'm searching for the rationale behind

> which includes the syntax being disabled for third-party image

At my company we have internal servers, trusted and everything, using the same IdP and session.

karlstolley commented:
@killermoehre you might want to join discussions of this on CZO; see [#issues > INLINE_IMAGE_PREVIEW applied inconsistently for local files @ 💬](https://chat.zulip.org/#narrow/channel/9-issues/topic/INLINE_IMAGE_PREVIEW.20applied.20inconsistently.20for.20local.20files/near/2482248) for some of the latest thinking about this.
